src/Microsoft.AspNetCore.Dispatcher/DispatcherValueEndpointSelector.cs

```diff
@@ -2,54 +2,262 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
 
 namespace Microsoft.AspNetCore.Dispatcher
 {
     public class DispatcherValueEndpointSelector : EndpointSelector
     {
-        public override Task SelectAsync(EndpointSelectorContext context)
+        private static readonly IReadOnlyList<SimpleEndpoint> EmptyEndpoints = Array.Empty<SimpleEndpoint>();
+
+        private EndpointSelectorContext _endpointSelectorContext;
+        private Cache _cache;
+
+        private Cache Current
+        {
+            get
+            {
+                var endpoints = _endpointSelectorContext.Endpoints;
+                var cache = Volatile.Read(ref _cache);
+
+                if (cache != null && cache.Version == endpoints.Version)
+                {
+                    return cache;
+                }
+
+                cache = new Cache(endpoints);
+                Volatile.Write(ref _cache, cache);
+                return cache;
+            }
+        }
```

Comment:
I'd like to see what this code looks like without the updating feature.

Can you instead add an `public virtual void Intialize(IEndpointCollection)` to the `EndpointSelector` class and do the initialization in there?

Call it from inside of `DispatcherBase`